File: peintre-nano/src/types/context-envelope.ts
/**
 * ContextEnvelope — contexte actif autoritatif (site, caisse, permissions effectives, etc.).
 * Schéma canonique : OpenAPI `recyclique` (backend). Piste A : mocks / stubs **structurellement alignés** jusqu'à Convergence 1.
 *
 * Champs marqués « stub » : à aligner sur le schéma OpenAPI canonique (`contracts/openapi/recyclique-api.yaml`) lorsque disponible.
 *
 * Story 25.11 (spike) — traçabilité checklist 25.7 / spec 25.4 §2–3 (IDs `CTX-*`) :
 * - `CTX-INV-2-1-*` / site explicite → `siteId` (ne pas inférer le site seul depuis l’UI pour du sensible).
 * - `CTX-INV-2-2-*` → `activeRegisterId` + cohérence avec `siteId` (vérité serveur).
 * - `CTX-INV-2-3-*` → `cashSessionId` + audit hors enveloppe UI.
 * - `CTX-SWITCH-3-1-*` (pas de stale client) → `context-envelope-freshness` + erreurs API explicites (`CONTEXT_STALE`, etc.).
 * ADR 25-2 / 25-3 : PIN kiosque, step-up, outbox — pas de champs supplémentaires imposés dans ce stub ; voir doc spike `_bmad-output/implementation-artifacts/2026-04-20-spike-25-11-contrats-enveloppe-contexte.md`.
 */

export interface EffectivePermissions {
  readonly permissionKeys: readonly string[];
}

/** État runtime côté UI (à rapprocher des codes erreur / statuts API backend). */
export type ContextEnvelopeRuntimeStatus = 'ok' | 'degraded' | 'forbidden';

export interface ContextEnvelopeStub {
  readonly schemaVersion: string;
  readonly siteId: string | null;
  readonly activeRegisterId: string | null;
  /**
   * Session caisse active lorsque le backend l'expose (`ContextEnvelope.context.cash_session_id` / exploitation).
   * Stub démo : peut rester null ; le parcours caisse 6.1 peut demander une saisie locale de secours (terrain).
   */
  readonly cashSessionId?: string | null;
  /**
   * Poste / workstation actif si le backend l’expose — sert au filtrage déclaratif `contexts_*` sur le marqueur `poste`.
   */
  readonly workstationId?: string | null;
  /**
   * Marqueurs de contexte actifs fournis par le backend (prioritaires sur la dérivation locale).
   * Ne pas inférer de permissions à partir de cette liste : consommation déclarative uniquement.
   */
  readonly contextMarkers?: readonly string[];
  readonly permissions: EffectivePermissions;
  /** Horodatage d'émission (ms depuis epoch) — alignement OpenAPI à confirmer. */
  readonly issuedAt: number;
  /**
   * Seuil de fraîcheur (ms) pour l'UI ; si absent, voir {@link MAX_CONTEXT_AGE_MS} dans `runtime/context-envelope-freshness.ts`.
   * Surcharge utile pour les tests.
   */
  readonly maxAgeMs?: number;
  /** Signaux explicites : ne pas inférer un contexte métier valide si `degraded` ou `forbidden`. */
  readonly runtimeStatus: ContextEnvelopeRuntimeStatus;
  /**
   * Aligné OpenAPI `presentation_labels` : résolution des `label_key` manifestes (navigation, etc.).
   * Présentation uniquement — ne participe pas aux décisions d’accès.
   */
  readonly presentationLabels?: Readonly<Record<string, string>>;
  /** Aligné OpenAPI `restriction_message` : message diagnostic UI (non autoritatif). */
  readonly restrictionMessage?: string | null;
  /**
   * Story 27.7 — intersection modules poste partagé (serveur).
   * Présent quand device + session opérateur ; utilisé pour projection navigation.
   */
  readonly effectiveModuleKeys?: readonly string[] | null;
}

File: recyclique/api/src/recyclic_api/schemas/context_envelope.py
"""Schémas ContextEnvelope minimal (Story 2.2) — alignés sur contracts/openapi/recyclique-api.yaml."""

from datetime import datetime
from enum import Enum
from typing import Dict, List

from pydantic import BaseModel, ConfigDict, Field


class ContextRuntimeState(str, Enum):
    """États runtime UI — noms figés contrat OpenAPI (spec 1.3 §4.2)."""

    ok = "ok"
    degraded = "degraded"
    forbidden = "forbidden"


class ExploitationContextIdsOut(BaseModel):
    """Identifiants de contexte visibles ; champs optionnels (brownfield)."""

    model_config = ConfigDict(extra="forbid")

    site_id: str | None = None
    cash_register_id: str | None = None
    cash_session_id: str | None = None
    reception_post_id: str | None = None
    # Epic 27.2 — poste partagé (null par défaut pour utilisateur web classique)
    device_id: str | None = Field(
        default=None,
        description="UUID canonique RegisteredDevice — distinct cash_register_id / reception_post_id.",
    )
    operator_user_id: str | None = Field(
        default=None,
        description="Opérateur PIN actif sur le poste (session serveur).",
    )
    module_key: str | None = Field(
        default=None,
        description="Module courant session opérateur ; intersection effective en story 27.7.",
    )
    override_active: bool | None = Field(
        default=None,
        description="Override SuperAdmin actif côté serveur (activation complète story 27.10).",
    )


class ContextEnvelopeResponse(BaseModel):
    """
    Enveloppe autoritaire serveur.
    `permission_keys` : sous-ensemble effectif minimal (Story 2.2) ; union rôles + groupes exhaustive en Story 2.3.
    """

    model_config = ConfigDict(extra="forbid")

    runtime_state: ContextRuntimeState
    context: ExploitationContextIdsOut | None = None
    permission_keys: List[str] = Field(default_factory=list)
    computed_at: datetime
    restriction_message: str | None = Field(
        default=None,
        description="Précision serveur pour degraded / forbidden (pas de logique métier côté client).",
    )
    presentation_labels: Dict[str, str] | None = Field(
        default=None,
        description="Map label_key CREOS → libellé affichable (Story 5.5) ; présentation uniquement.",
    )
    effective_module_keys: List[str] | None = Field(
        default=None,
        description=(
            "Story 27.7 — intersection serveur site × allowlist poste × permissions opérateur. "
            "Présent uniquement quand device_id + operator_user_id actifs ; sinon absent ou []."
        ),
    )
